src/materialized/src/lib.rs

```diff
@@ -320,3 +333,57 @@ impl Server {
         self.local_addr
     }
 }
+
+// Runs fetch_latest_version in a backoff loop until it succeeds once, prints a
+// warning if there is another version, then returns.
+fn check_version_loop(telemetry_url: String, cluster_id: String) {
+    let mut backoff = Duration::from_secs(5);
+    loop {
+        let latest_version = match fetch_latest_version(&telemetry_url, &cluster_id) {
+            Ok(version) => version,
+            Err(err) => {
+                backoff *= 2;
+                info!(
+                    "could not fetch latest version: {}; retrying in {:?}",
+                    err, backoff
+                );
+                thread::sleep(backoff);
+                continue;
+            }
+        };
+
+        if latest_version != version() {
+            warn!(
+                "A new version of materialized is available: {}.",
+                latest_version
+            );
+        }
+        return;
+    }
+}
+
+fn fetch_latest_version(telemetry_url: &str, cluster_id: &str) -> anyhow::Result<String> {
+    let version_url = format!("{}/api/v1/version/{}", telemetry_url, cluster_id);
+    let version_request = V1VersionRequest { version: version() };
+
+    let resp = reqwest::blocking::Client::new()
+        .post(&version_url)
```

Comment:
Does this correctly handle the HTTP "moved temporarily/permanently" responses? We probably want to reserve the right to change the URL in the future.

Reply:
Yes, reqwest will follow 10 redirects by default.